### PR TITLE
feat: rts from committed changes

### PR DESCRIPTION
**Pull request by guotin**
Here's a version that can check tests from committed changes with `--rts-from-commit=[init commithash]`. If it's not provided, only changes in the working directory are checked. A slight problem is that new tests are discovered independently from the workdir/committed option and it would require some extra checking to get things perfectly right. 

Wrote some testcases for various situations. Also noticed that I created really flaky tests in the last PR because `__pycache__` files in the testrepo didn't update between runs. I think I fixed it by clearing them before each test.

### CODE REVIEW

File: pytest_rts/utils/git.py [hunk 1]
```diff
@@ -1,26 +1,38 @@
...
     """Get changed files in git working directory"""
     return repo.repo.git.diff("--name-only").split()
 
 
-def get_file_diff_data_current(repo: GitRepository, filename: str) -> str:
+def get_changed_files_between_commits(
+    repo: GitRepository, commithash1: str, commithash2: str
+) -> List[str]:
+    """Get changed files between two commits"""
+    return repo.repo.git.diff("--name-only", commithash1, commithash2).split()
```
Comment (matveypashkovskiy) [inline pytest_rts/utils/git.py:23]: If I'm not mistaken, in git diff command last commit hash could be omitted and that will automatically do diff between the first commit hash and HEAD. Please check.

File: pytest_rts/utils/git.py [hunk 2]
```diff
@@ -1,26 +1,38 @@
...
-def get_file_diff_dict_current(repo: GitRepository, files: List[str]) -> Dict[str, str]:
-    """Returns a dictionary with file id as key and git diff as value"""
-    return {
-        file_path: get_file_diff_data_current(repo, file_path) for file_path in files
-    }
+def get_file_diff_data_between_commits(
+    repo: GitRepository, file_path: str, commithash1: str, commithash2: str
+) -> str:
+    """Get git diff for a file from changes between two commits"""
+    return repo.repo.git.diff("-U0", commithash1, commithash2, "--", file_path)
```
Comment (matveypashkovskiy) [inline pytest_rts/utils/git.py:35]: same here: maybe we can omit `commithash2` and thus `get_current_head_hash` won't be needed

File: pytest_rts/tests/test_e2e.py [hunk 3]
```diff
@@ -39,7 +113,11 @@ def test_only_new_functions_are_ran(testdir: Testdir) -> None:
     os.rename(".coverage", COV_FILE)
```
Comment (matveypashkovskiy) [inline pytest_rts/tests/test_e2e.py:113]: It looks like lines above together with `cleanup_caches(str(testdir.tmpdir))` are called in every test and thus are good candidates for [`setup_module()`](https://docs.pytest.org/en/stable/xunit_setup.html) method. Isn't it?

**Review by matveypashkovskiy — CHANGES_REQUESTED** [review]

File: pytest_rts/tests/test_e2e.py (hunk 3 above)
Comment (guotin) [reply pytest_rts/tests/test_e2e.py:105]: I used a pytest fixture to do something like this

File: pytest_rts/utils/git.py (hunk 1 above)
Comment (guotin) [reply pytest_rts/utils/git.py]: Giving only the first commithash results in both workdir and committed changes to be displayed in the diff so I used that.

File: pytest_rts/plugin.py [hunk 4]
```diff
@@ -36,6 +46,17 @@ def pytest_configure(config: Config) -> None:
...
+    if not commit_exists(config.option.rts_from_commit):
+        pytest.exit("pytest-rts: Given commithash does not exist", 2)
+
     existing_tests = get_existing_tests(config.option.rts_coverage_db)
-    workdir_tests = get_tests_current(config.option.rts_coverage_db)
-    config.pluginmanager.register(RunnerPlugin(existing_tests, workdir_tests))
+    committed_tests = get_tests_committed(
+        config.option.rts_from_commit, config.option.rts_coverage_db
+    )
+    config.pluginmanager.register(RunnerPlugin(existing_tests, committed_tests))
```
[the remark this answers is not in the document]
Comment (guotin) [reply pytest_rts/plugin.py]: I think it should now work as you described. The given sha is verified to exist and then checking for both workdir + committed is done.

File: README.md [hunk 5]
```diff
@@ -26,12 +26,23 @@ relative_files = True
...
+
 1. Install `pytest-rts` with `pip install pytest-rts`
 2. Create a branch `git checkout -b feat/new-feature`
 3. Make changes in your code
 4. Run the tool with `pytest --rts --rts-coverage-db=[path to database]`
 
-As a result only tests related to changes in working directory and branch will be executed.
+As a result only tests related to changes in working directory will be executed.
+
+#### Tests from working directory + committed changes
```
Comment (matveypashkovskiy) [inline README.md:38]: Let's have the title starting the same way as previous: "Tests from changes in Git working directory + committed changes" to make it clear that it is the same as previous with some extra.

File: pytest_rts/utils/git.py [hunk 6]
```diff
@@ -1,26 +1,40 @@
...
 
 from pydriller import GitRepository
 
 
-def get_changed_files_current(repo: GitRepository) -> List[str]:
+def commit_exists(commithash: str, repo: GitRepository) -> bool:
+    """Check if a given commithash exists in the Git repository"""
+    if not commithash:
+        return False
+    return commithash in [commit.hash for commit in repo.get_list_commits()]
```
Comment (matveypashkovskiy) [inline pytest_rts/utils/git.py:13]: * it is better to use `repo.get_commit(commithash)` here
* print warning if `commithash` is provided but doesn't exist in history

File: pytest_rts/utils/common.py [hunk 7]
```diff
@@ -43,9 +46,28 @@ def get_existing_tests(coverage_file_path: str) -> Set[str]:
...
+    commithash_to_compare: str, coverage_file_path: str
 ) -> Set[str]:
-    """Returns the test set from given Git diffs"""
+    """Returns the test set from Git changes.
+    If the provided commithash does not exist in the repo,
+    then only changes in the git working directory are considered.
+    Otherwise the commithash is compared to the current working copy.
+    """
+    repo = get_git_repo()
+    if not commit_exists(commithash_to_compare, repo):
```
Comment (matveypashkovskiy) [inline pytest_rts/utils/common.py:57]: Inverted condition is more readable: compare `if commit_exists()` with `if not commit_exists()`

**Review by matveypashkovskiy — CHANGES_REQUESTED** [review]

File: pytest_rts/tests/helper_project/.gitignore [hunk 8]
```diff
@@ -1,3 +1,6 @@
 .coverage
-rts-coverage.1234
+rts-coverage.db
 *__pycache__*
+runpytest-*
+stderr
+stdout
```
Comment (alexey-vyskubov) [inline pytest_rts/tests/helper_project/.gitignore:6]: What are stderr and stdout here?

Comment (guotin) [reply pytest_rts/tests/helper_project/.gitignore:6]: Files produced by the `pytester` plugin. I thought ignoring them would make debugging test cases for running working directory changes easier.

**Pull request merged**